clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,190 @@
+            let mut suggestions: Vec<(Span, String)> = vec![];
+            let mut message = "consider using `if..let` syntax instead of indexing".to_string();
+            span_lint_and_then(
+                cx,
+                UNNECESSARY_INDEXING,
+                expr.span,
+                "condition can be simplified with `if..let` syntax",
+                |diag| {
+                    if let Some(first_local) = r.first_local
+                        && let Some(name) = first_local.pat.simple_ident().map(|ident| ident.name)
+                    {
+                        suggestions.push((
+                            if_expr.cond.span,
+                            format!(
+                                "let Some({}{name}) = {receiver}.first()",
+                                // if we arent borrowing anything then we can pass a reference here for correctness
+                                if r.extra_exprs_borrow.is_empty() { "&" } else { "" },
+                            ),
+                        ));
+                        suggestions.push((first_local.span, String::new()));
+
+                        if !r.extra_exprs_borrow.is_empty() {
+                            suggestions.extend(
+                                r.extra_exprs_borrow
+                                    .iter()
+                                    .chain(r.extra_exprs_copy.iter())
+                                    .map(|x| (x.span, name.to_string())),
+                            );
+
+                            message.push_str(", and replacing indexing expression(s) with the value in `Some` variant");
+                        } else if !r.extra_exprs_copy.is_empty() {
+                            suggestions.extend(r.extra_exprs_copy.iter().map(|x| (x.span, name.to_string())));
+                        }
+                    } else if r.extra_exprs_borrow.is_empty() {
+                        suggestions.push((if_expr.cond.span, format!("let Some(&element) = {receiver}.first()")));
+                        suggestions.extend(r.extra_exprs_copy.iter().map(|x| (x.span, "element".to_owned())));
+                    } else {
+                        suggestions.push((if_expr.cond.span, format!("let Some(element) = {receiver}.first()")));
+                        suggestions.extend(
+                            r.extra_exprs_borrow
+                                .iter()
+                                .chain(r.extra_exprs_copy.iter())
+                                .map(|x| (x.span, "element".to_owned())),
+                        );
+                    }
+
+                    diag.multipart_suggestion(message, suggestions, Applicability::MaybeIncorrect);
+                },
+            );
+        }
+    }
+}
+
+struct IndexCheckResult<'a> {
+    // span of the receiver for the index operation, only Some in the event the indexing is via a direct primitive
+    index_receiver_span: Option<Span>,
```

Comment:
The `Option` here can be removed since `process_indexing` already returns an `Option<IndexCheckResult>`